src/cpprealm/flex_sync.hpp

```diff
@@ -0,0 +1,172 @@
+    }
+
+    std::unique_ptr<sync::Subscription> m_subscription;
+};
+
+// A MutableSyncSubscriptionSet represents a single query that may be OR'd with other queries on the same object class to be
+// send to the server in a QUERY or IDENT message.
+struct MutableSyncSubscriptionSet {
+public:
+    // Inserts a new subscription into the set if one does not exist already.
+    // If the `query_fn` parameter is left empty, the subscription will sync *all* objects
+    // for the templated class type.
+    template <type_info::ObjectPersistable T>
+    void add(const std::string& name, std::optional<std::function<rbool(T&)>> query_fn = std::nullopt) {
+        auto schema = *m_realm->schema().find(T::schema::name);
+        auto& group = m_realm->read_group();
+        auto table_ref = group.get_table(schema.table_key);
+        auto builder = Query(table_ref);
+
+        if (query_fn) {
+            auto q = query<T>(builder, std::move(schema));
+            auto full_query = (*query_fn)(q).q;
+            m_subscription_set.insert_or_assign(name, full_query);
+        } else {
+            m_subscription_set.insert_or_assign(name, builder);
+        }
+    }
+
+    // Removes a subscription for a given name. Will throw if subscription does
+    // not exist.
+    void remove(const std::string& name) {
+        auto it = m_subscription_set.find(name);
+        if (it != m_subscription_set.end()) {
+            m_subscription_set.erase(it);
+            return;
+        }
+        throw std::logic_error("Subscription cannot be found");
+    }
+
+    // Finds a subscription for a given name. Will return `std::nullopt` is subscription does
+    // not exist.
+    std::optional<SyncSubscription> find(const std::string& name) {
+        auto it = m_subscription_set.find(name);
+        if (it != m_subscription_set.end()) {
+            return SyncSubscription {std::make_unique<realm::sync::Subscription>(*it)};
+        }
+        return std::nullopt;
+    }
+
+    // Updates a subscription for a given name.
+    // Will throw if subscription does not exist.
+    // If the `query_fn` parameter is left empty, the subscription will sync *all* objects
+    // for the templated class type.
+    template <type_info::ObjectPersistable T>
+    void update_subscription(const std::string& name, std::optional<std::function<rbool(T&)>> query_fn = std::nullopt) {
+        remove(name);
+        add(name, query_fn);
+    }
+
+    // Removes all subscriptions.
+    void clear() {
+        m_subscription_set.clear();
+    }
+
+    MutableSyncSubscriptionSet(sync::MutableSubscriptionSet subscription_set, SharedRealm realm) :
+    m_subscription_set(subscription_set)
+    , m_realm(realm) {}
+private:
+    sync::MutableSubscriptionSet m_subscription_set;
+    SharedRealm m_realm;
+
+    friend struct SyncSubscriptionSet;
+    sync::MutableSubscriptionSet get_subscription_set()
+    {
+        return m_subscription_set;
+    }
+};
+
+struct SyncSubscriptionSet {
```

Comment:
Should be lower snake case. Though this isn't really cpp naming convention, we currently need the layer of naming distinction with ObjStore/Core.

Reply:
can we not use distinguish them through namespaces instead of snake case?